From 34caaa33c0628c75ca5d334bc4924424d4674732 Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Thu, 11 Sep 2014 12:00:42 -0400
Subject: [ethernet] mlx4: Adjust port number in qp_attach wrapper when detaching

Message-id: <c74b292e5ffb86a111d1b7f682dd7da02688c480.1410425016.git.avadai@redhat.com>
Patchwork-id: 93648
O-Subject: [Patch RHEL7.1 V1 106/197] net/mlx4_core: Adjust port number in qp_attach wrapper when detaching
Bugzilla: 1107617 1107618 1107619 1107620
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Jack Morgenstein <jackm@dev.mellanox.co.il>

When using single ported VFs and the VF is using port 2, we need
to adjust the port accordingly (change it from 1 to 2).

Fixes: 449fc48 ('net/mlx4: Adapt code for N-Port VF')
Signed-off-by: Matan Barak <matanb@mellanox.com>
Signed-off-by: Jack Morgenstein <jackm@mellanox.com>
Signed-off-by: Or Gerlitz <ogerlitz@mellanox.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit 531d9014d5c870fdf493e626c4b4e448273cb616)
Signed-off-by: Amir Vadai <avadai@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/mellanox/mlx4/resource_tracker.c b/drivers/net/ethernet/mellanox/mlx4/resource_tracker.c
index 3b5f53e..1c3fdd4 100644
--- a/drivers/net/ethernet/mellanox/mlx4/resource_tracker.c
+++ b/drivers/net/ethernet/mellanox/mlx4/resource_tracker.c
@@ -3733,6 +3733,25 @@ static int qp_detach(struct mlx4_dev *dev, struct mlx4_qp *qp,
  }
 }
 
+static int mlx4_adjust_port(struct mlx4_dev *dev, int slave,
+       u8 *gid, enum mlx4_protocol prot)
+{
+ int real_port;
+
+ if (prot != MLX4_PROT_ETH)
+  return 0;
+
+ if (dev->caps.steering_mode == MLX4_STEERING_MODE_B0 ||
+     dev->caps.steering_mode == MLX4_STEERING_MODE_DEVICE_MANAGED) {
+  real_port = mlx4_slave_convert_port(dev, slave, gid[5]);
+  if (real_port < 0)
+   return -EINVAL;
+  gid[5] = real_port;
+ }
+
+ return 0;
+}
+
 int mlx4_QP_ATTACH_wrapper(struct mlx4_dev *dev, int slave,
           struct mlx4_vhcr *vhcr,
           struct mlx4_cmd_mailbox *inbox,
@@ -3768,6 +3787,10 @@ int mlx4_QP_ATTACH_wrapper(struct mlx4_dev *dev, int slave,
   if (err)
    goto ex_detach;
  } else {
+  err = mlx4_adjust_port(dev, slave, gid, prot);
+  if (err)
+   goto ex_put;
+
   err = rem_mcg_res(dev, slave, rqp, gid, prot, type, &reg_id);
   if (err)
    goto ex_put;
-- 
1.7.1